catsoop/lti.py

```diff
@@ -0,0 +1,297 @@
+    '''
+    session_data = context["cs_session_data"]
+    logging = context["csm_cslog"]
+    LOGGER.info("[lti.get_or_create_user] uname=%s, email=%s, name=%s" % (uname, email, name))
+
+    login_info = logging.most_recent("_logininfo", [], uname, {})
+    if not login_info:	# account doesn't exist, create
+
+        LOGGER.info("[lti.get_or_create_user] uname=%s unknown username -> creating new account")
+        passwd = ''.join([random.choice(string.ascii_letters + string.digits) for n in range(32)])
+        hash_iterations = context.get("cs_password_hash_iterations", 500000)
+        salt = get_new_password_salt()
+        phash = compute_password_hash(context, passwd, salt, hash_iterations)
+        confirmed = True
+        
+        uinfo = {
+            "password_salt": salt,
+            "password_hash": phash,
+            "email": email,
+            "name": name,
+            "confirmed": confirmed,
+        }
+        logging.overwrite_log("_logininfo", [], uname, uinfo)
+        login_info = logging.most_recent("_logininfo", [], uname, {})
+
+    LOGGER.info("[lti.get_or_create_user] login_info=%s" % login_info)
+
+    if login_info is None:	# account creation failed
+        msg = "[lti.get_or_create_user] failed to create user %s" % uinfo
+        LOGGER.error(msg)
+        raise Exception(msg)
+
+    info = {"username": uname, "name": name, "email": email,
+            "is_lti_user": True}		# see preload.py in course data; forces authorization
+    session_data.update(info)
+
+    LOGGER.info("[lti.get_or_create_user] login_info = %s" % login_info)
+
+    user_info = auth.get_logged_in_user(context)
+    LOGGER.info("[lti.get_or_create_user] user_info=%s" % user_info)
+    context["cs_user_info"] = user_info
+    context["cs_username"] = str(user_info.get("username", None))
+
+    session.set_session_data(context, context["cs_sid"], session_data)	# save session data
+
+    return login_info
+
+
+def serve_lti(context, path_info, environment, params, dispatch_main):
+    '''
+    context: (dict) catsoop global context
+    path_info: (list) URL path components
+    environment: (dict-like) web server data, such as form input
+    dispatch_main: (proc) call this with environment to dispatch to render URL
+    '''
+    if not 'cs_lti_config' in context:
+        msg = "[lti] LTI not configured - missing cs_lti_config in config.py"
+        LOGGER.error(msg)
+        raise Exception(msg)
+
+    LOGGER.info("[lti] parameters=%s" % params)
+    lti_action = path_info[0]
+    LOGGER.info("[lti] lti_action=%s, path_info=%s" % (lti_action, path_info))
+
+    session = context['cs_session_data']
+    l4c = lti4cs(context, session, {}, {})
+    lti_ok = l4c.verify_request(params, environment)
+    if not lti_ok:
+        msg = "LTI verification failed"
+    else:
+        lti_data = session["lti_data"]
+        uname = "lti_%s" % lti_data.get("lis_person_sourcedid", lti_data['user_id'])
```

Comment:
I'm wondering if there is a way to avoid a potential overlap between these usernames and other usernames that might exist on the system.  If a CAT-SOOP instance is being both as an LTI provider and as a standalone instance, for example, then it seems like the user information could get confused (e.g., if a user called `lti_foo` existed in CAT-SOOP, and this function was called where `lti_data['user_id']` was `foo`, then those two accounts would be confused).

Or am I misunderstanding things, and maybe `lti_data['user_id']` is not a username _per se_, but a long unique identifier of some kind?  In that case, I'm not worried.

Reply:
Yes - there indeed could be a collision;  `lti_data['user_id']` is indeed numeric, but `lis_person_sourcedid` can be a username string (sometimes it's a uuid).

How about allowing the `lti_` username prefix to be configurable?  This way each catsoop installation could determine their best way to avoid a collision.

Reply:
I think allowing configuration of the prefix seems good.

Along those lines, is there a reason to use `"lis_person_sourcedid"`, as opposed to using `lti_data['user_id']` always, since (I think) the expectation is that these accounts won't be used by people logging into CAT-SOOP directly?  I think it would make collisions far less of a concern.

Reply:
You have admin permissions because I should be mapping LTI roles (from MITx) properly to CAT-SOOP roles.  You're an admin on the MITx site so you should see an admin view on CAT-SOOP, but it's actually not yet being done properly for other roles.

Reply:
configurable prefix added: https://github.com/ichuang/catsoop/pull/2/commits/88ee8ea969fd37dda2a2f79c90c9e724474e0b10 (in  https://github.com/ichuang/catsoop/pull/2)